drivers/flash/flash_stm32h7x.c

```diff
@@ -469,21 +549,56 @@ flash_stm32h7_get_parameters(const struct device *dev)
 	return &flash_stm32h7_parameters;
 }
 
-static const struct flash_pages_layout stm32h7_flash_layout[] = {
-	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
-#ifdef DUAL_BANK
-	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
-#endif
-};
 
 void flash_stm32_page_layout(const struct device *dev,
 			     const struct flash_pages_layout **layout,
 			     size_t *layout_size)
 {
 	ARG_UNUSED(dev);
 
+#if defined(DUAL_BANK)
+	static struct flash_pages_layout stm32h7_flash_layout[3];
+
+	if (REAL_FLASH_SIZE < KB(STM32H7_SERIES_MAX_FLASH)) {
+		if (stm32h7_flash_layout[0].pages_count == 0) {
+			/* Bank1 */
+			stm32h7_flash_layout[0].pages_count = SECTOR_PER_BANK;
+			stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
+			/*
+			 * Dummy page corresponding to discontinuity
+			 * between bank1/2
+			 */
+			stm32h7_flash_layout[1].pages_count = 1;
+			stm32h7_flash_layout[1].pages_size = BANK2_OFFSET
+					- (SECTOR_PER_BANK * FLASH_SECTOR_SIZE);
+			/* Bank2 */
+			stm32h7_flash_layout[2].pages_count = SECTOR_PER_BANK;
+			stm32h7_flash_layout[2].pages_size = FLASH_SECTOR_SIZE;
+		}
+		*layout = stm32h7_flash_layout;
+		*layout_size = ARRAY_SIZE(stm32h7_flash_layout);
+	} else {
+		static struct flash_pages_layout stm32h7_flash_layout[1];
+
+		if (stm32h7_flash_layout[0].pages_count == 0) {
+			stm32h7_flash_layout[0].pages_count =
+				REAL_FLASH_SIZE / FLASH_SECTOR_SIZE;
+			stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
+		*layout = stm32h7_flash_layout;
+		*layout_size = 1;
+		}
+	}
+#else
+	static struct flash_pages_layout stm32h7_flash_layout[1];
+
+	if (stm32h7_flash_layout[0].pages_count == 0) {
+		stm32h7_flash_layout[0].pages_count =
+				REAL_FLASH_SIZE / FLASH_SECTOR_SIZE;
+		stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
+	}
```

Comment:
No way to factorize these 2 parts ?

Reply:
No due to array declaration (refatored) and I think it would deserve readability.